Implement the evalsha and script command

git-hulk opened the pull request:
This PR implements the `evalsha` and `script`(except debug and kill subcommand) command, `script kill` and Lua time limit would be supported in the next one to keep the PR simple.

git-hulk committed:
Implement the evalsha and script command

git-hulk committed:
Support to propagate and replay the script flush/load command

git-hulk committed:
Add test cases

git-hulk committed:
Export the cached scripts count and estimated cached scripts memory

ShooterIT commented on src/server.h:
```diff
@@ -158,6 +170,9 @@ class Server {
 
   lua_State *lua_;
+  std::mutex lua_scripts_mu_;
+  std::unordered_map<std::string, std::string> lua_scripts_;
```
Why don't we store the lua scripts into rocksdb instead of memory? it may cost much more memory.

And we also regard lua scripts as key value pairs in propagate CF, so we don't handle it in replication.

git-hulk replied:
As you say, it's a good idea to persist those scripts on rocksdb.

git-hulk committed:
Persist lua scripts into the propagate column family

Redis regards the lua scripts as memory cache which would
lost after restarting, but we prefer persisting into the
disk which also can save some memory usage. The only side
effect was all history scripts would be stored, but we think
it's ok since scripts shouldn't be too many.

git-hulk committed:
Fix should NOT flush scripts on slave

git-hulk committed:
Add the 'f_' prefix to the script key and rename 'command' to 'lua'

ShooterIT commented on src/scripting.cc:
```diff
@@ -676,7 +784,7 @@ Status createFunction(lua_State *lua, const std::string &body, std::string *sha)
                   "Error running script (new function): " + errMsg + "\n");
   }
-  *sha = funcdef.substr(2);
+  srv->ScriptSet(*sha, body);
```
please add comments here, such as `store lua script into xx cf of rocksdb and will auto propagate it`

ShooterIT commented on src/server.cc:
```diff
@@ -1222,3 +1222,76 @@ Status Server::LookupAndCreateCommand(const std::string &cmd_name,
+
+void Server::ScriptFlush() {
+  auto cf = storage_->GetCFHandle(Engine::kPropagateColumnFamilyName);
+  storage_->DeleteAll(rocksdb::WriteOptions(), cf);
```
we shouldn't delete all, we just want to delete keys that have `f_` prefix

git-hulk replied:
good catch, it's my carelessness

ShooterIT commented on src/server.cc:
```diff
@@ -1222,3 +1222,76 @@ Status Server::LookupAndCreateCommand(const std::string &cmd_name,
+}
+
+void Server::ScriptSet(const std::string &sha, const std::string &body) {
+  std::string funcname = "f_" + sha;
```
maybe we create one const variable(`lua_script_prefix`) instead of `f_` here.

i think `lua_f_` is better than `f_`, WDYT?

git-hulk replied:
both were ok to me, so pick the `lua_f`

ShooterIT commented on src/storage.cc:
```diff
@@ -30,6 +30,9 @@ const char *kPubSubColumnFamilyName = "pubsub";
 const char *kSubkeyColumnFamilyName = "default";
+const char *kPropagateColumnFamilyName = "propagate";
+
+const char *kPropagateTypeKey = "lua";
```
kPropagateTypeKey -> kPropagateLua？

ShooterIT commented on src/replication.cc:
```diff
@@ -917,11 +917,20 @@ rocksdb::Status ReplicationThread::ParseWriteBatch(const std::string &batch_stri
+      srv_->PublishMessage(write_batch_handler.Key(), write_batch_handler.Value());
+      break;
+    case kBatchTypePropagate:
+      if (write_batch_handler.Key() == "command") {
```
still `command`？

ShooterIT commented on tests/tcl/tests/unit/scripting.tcl:
```diff
@@ -42,21 +42,40 @@ start_server {tags {"scripting"}} {
+    test {EVALSHA - Do we get an error on non defined SHA1?} {
+        catch {r evalsha ffd632c7d33e571e9f24556ebed26c3479a87130 0} e
+        set _ $e
+    } {ERR NOSCRIPT*}
```
i think, in lua test, you should add mater-slaves replication test case.

git-hulk replied:
There's a little tricky on the script command but at least we have the test case.

git-hulk committed:
Fix issues found by CR

git-hulk committed:
Add scripting repl test case

git-hulk committed:
Fix wrong status in tcl util

git-hulk committed:
Fix should use lua_ltostring to get the lua argument

ShooterIT commented on tests/tcl/tests/support/util.tcl:
```diff
@@ -95,7 +95,7 @@ proc wait_for_sync r {
 
 proc wait_for_ofs_sync {r1 r2} {
     wait_for_condition 50 100 {
-        [status $r1 master_repl_offset] eq [status $r2 master_repl_offset]
```
i think it is time we should support `master_repl_offset` in INFO command output

git-hulk replied:
You're right, how about improving this in the next PR?

ShooterIT replied:
OK

ShooterIT commented on tests/tcl/tests/unit/scripting.tcl:
```diff
@@ -252,4 +271,104 @@ start_server {tags {"scripting"}} {
+            } else {
+                fail "Replication not started."
+            }
+        }
```
```suggestion
        $slave slaveof $master_host $master_port
        wait_for_sync $slave
```

ShooterIT commented on docs/support-commands.md:
```diff
@@ -148,6 +148,13 @@
+| --------- | ---------------- | ---- |
+| eval      | √                |      |
+| evalsha   | √                |      |
+| script    | √                |      |
```
script kill and debug subcommand are not supported

ShooterIT commented on src/compact_filter.h:
```diff
@@ -71,6 +71,26 @@ class SubKeyFilterFactory : public rocksdb::CompactionFilterFactory {
   Engine::Storage *stor_ = nullptr;
 };
 
+class TransitFilter : public rocksdb::CompactionFilter {
```
i accept 'propagate' name, so i think here we should also can use PropagateFilter, WDYT?
especially, you use `CompactTransitFiles` for `pubsub` and `propagate` CF, it is confused.

git-hulk replied:
good catch.

Initially, I use the same compact filter(pubsub compact filter) between propagate and pubsub, so renamed it as `transit` compact filter, but didn't revert this name after splitting.

ShooterIT commented on src/compact_filter.h:
```diff
@@ -71,6 +71,26 @@ class SubKeyFilterFactory : public rocksdb::CompactionFilterFactory {
+  const char *Name() const override { return "TransitFilter"; }
+  bool Filter(int level, const Slice &key, const Slice &value,
+              std::string *new_value, bool *modified) const override {
+    // only filter the propagating command here
```
```suggestion
    // We propagate Lua commands which don't store data,
    // just in order to implement updating Lua state.
```
How about above comments

ShooterIT commented on src/redis_cmd.cc:
```diff
@@ -4473,6 +4524,8 @@ CommandAttributes redisCommandTable[] = {
 
     ADD_CMD("eval", -3, "exclusive write no-script", 0, 0, 0, CommandEval),
+    ADD_CMD("evalsha", -3, "exclusive write no-script", 0, 0, 0, CommandEvalSHA),
+    ADD_CMD("script", -2, "exclusive read-only no-script", 0, 0, 0, CommandScript),
```
```suggestion
    ADD_CMD("script", -2, "exclusive no-script", 0, 0, 0, CommandScript),
```
remove `read-only` attribute ?

git-hulk replied:
we allow executing `script exists` command on slave, so it should perform as `read-only`

ShooterIT replied:
if no `read-only` attribute, we don't consider it as write command, right?

git-hulk replied:
actually, some subcommands weren't, but it's ok that we only allow the `exists` subcommand when the role was a slave.

ShooterIT commented on tests/tcl/tests/tmp/.gitignore:
```diff
@@ -1 +0,0 @@
-*
```
Why do you delete this file

git-hulk replied:
my mistake

ShooterIT commented on src/storage.h:
```diff
@@ -21,13 +21,19 @@ enum ColumnFamilyID{
 extern const char *kSubkeyColumnFamilyName;
+extern const char *kPropagateColumnFamilyName;
+
+extern const char *kPropagateTypeLua;
```
```suggestion
extern const char *kPropagateLuaCommand;
```
this is a key name, i think `type` word is not clear.

`key` | `value`
lua -> script load
lua -> script flush

git-hulk replied:
agree

ShooterIT commented on src/server.cc:
```diff
@@ -1222,3 +1222,76 @@ Status Server::LookupAndCreateCommand(const std::string &cmd_name,
+  return Status::OK();
+}
+
+Status Server::Propagate(const std::string &type, const std::vector<std::string> &tokens) {
```
```suggestion
// Generally, we store data into rocksdb and just replicate WAL instead of propagating
// commands. But sometimes, we need to update inner states or do special operations
// for specific commands, such as `script flush`.
// channel: we put the same function commands into one channel to handle uniformly 
// tokens: the serialized commands
Status Server::Propagate(const std::string &channel, const std::vector<std::string> &tokens) {
```

ShooterIT replied:
i changed parameter `const std::string &type` into `const std::string &channel`, and in comments, i prefer channel, WDYT?

git-hulk replied:
Fixed

ShooterIT commented on src/server.cc:
```diff
@@ -1222,3 +1222,76 @@ Status Server::LookupAndCreateCommand(const std::string &cmd_name,
+    ScriptReset();
+    return Status::OK();
+  } else if (subcommand == "load" && tokens.size() == 3) {
+    // no need to replay the script load command, eval/evalsha would load the script from
```
so i think we don't need to propagate this subcommand.

git-hulk replied:
this command wasn't propagated now, removed this condition

ShooterIT commented on src/server.cc:
```diff
@@ -1222,3 +1222,76 @@ Status Server::LookupAndCreateCommand(const std::string &cmd_name,
+  return Status(Status::NotOK, "Unknown SCRIPT subcommand or wrong # of args");
+}
+
+Status Server::ReplayCommand(const std::vector<std::string> &tokens) {
```
i don't think `ReplayCommand` is a good name. it seems loading data, we aim to execute propagated commands on slaves. how about `ExecPropagatedCommand`, `ExecPropagatedLuaCommand`

git-hulk replied:
`ExecPropagatedCommand ` is better than ReplayCommand, but how about `ExecPropagatedLuaCommand` to ExecPropagatedScriptCommand` should be more preciously?

ShooterIT replied:
yes, `ExecPropagatedScriptCommand` is better

ShooterIT reviewed (commented):
A lot of comments, maybe it makes no sense, please check
i think we should add more comments for some special operations, otherwise, it is hard to understand.

git-hulk committed:
Fix CR issues

git-hulk commented:
> A lot of comments, maybe it makes no sense, please check i think we should add more comments for some special operations, otherwise, it is hard to understand.

most advice totally made sense and was great, I have fixed them but some comments. You can have a look again.

ShooterIT commented on src/storage.h:
```diff
@@ -21,13 +21,19 @@ enum ColumnFamilyID{
 extern const char *kSubkeyColumnFamilyName;
+extern const char *kPropagateColumnFamilyName;
+
+extern const char *kPropagateLuaCommand;
```
my bad, `kPropagateScriptCommand` seems better?

git-hulk replied:
oh no, it's my fault, it should be `kPropagateScriptCommand `

git-hulk committed:
Fix some incorrect variable names

git-hulk committed:
Update src/redis_cmd.cc

Co-authored-by: Wang Yuan <wangyuancode@163.com>

ShooterIT reviewed (approved):
Cool job,  cheers!
